vcd/resource_vcd_external_network.go

```diff
@@ -0,0 +1,334 @@
+package vcd
+
+import (
+	"fmt"
+	"log"
+	"net"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	types "github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+func resourceVcdExternalNetwork() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVcdExternalNetworkCreate,
+		Delete: resourceVcdExternalNetworkDelete,
+		Read:   resourceVcdExternalNetworkRead,
+		Update: resourceVcdExternalNetworkUpdate,
+
+		Schema: map[string]*schema.Schema{
+			"org": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Organization to create the external network in",
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: false,
+			},
+			"ip_scope": &schema.Schema{
+				Type:     schema.TypeList,
+				Required: true,
+				ForceNew: false,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"is_inherited": &schema.Schema{
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"gateway": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: singleIP(),
+						},
+						"netmask": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: singleIP(),
+						},
+						"dns1": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: singleIP(),
+						},
+						"dns2": &schema.Schema{
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: singleIP(),
+						},
+						"ip_range": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"start": &schema.Schema{
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: singleIP(),
+									},
+									"end": &schema.Schema{
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: singleIP(),
+									},
+								},
+							},
+						},
+					},
+				},
+				Description: "",
+			},
+			"vim_port_group": &schema.Schema{
```

Comment:
We should simplify this, I think name should be enough and everything else we can query.

Reply:
Ok, I know how to use the query endpoint to do a lookup via the name, but how would I get the `MoRef` and `VimObjectType`?

Reply:
I think we should follow UI where you choose port group and everything else will be automatically filled. PortGroup can be queried with MoRef and I think saw that VimObjectType is returned as "Network"

Reply:
Normally I'd use the browser tools to inspect the network calls being made, but flash apps use a protocol on top of http that I'm not familiar with, so it's difficult for me to understand what the UI is doing.

I made a request to `/api/admin/extension/vimServer/<id>/networks` and got
```
<vmext:VimObjectRefList ...>
    <vmext:VimObjectRefs>
        <vmext:VimObjectRef>
            <vmext:VimServerRef ... />
            <vmext:MoRef>dvportgroup-XXXX</vmext:MoRef>
            <vmext:VimObjectType>DV_PORTGROUP</vmext:VimObjectType>
        </vmext:VimObjectRef>
        ...
    </vmext:VimObjectRefs>
</vmext:VimObjectRefList>
```
I see a few dozen `<VimObjectRef>`s in the response.  It looks like if I had the `MoRef` as an input, I could ascertain the `VimObjectType`.  Is it ok for me to keep that input around, or do you see another way?

Reply:
I tried a little and if I query `portgroup`, I can select all info using port group name, e.g. of query of all port groups in my env, which can be improved using additional filter by name:

```
2019/04/11 08:11:04 --------------------------------------------------------------------------------
2019/04/11 08:11:04 Request caller: runtime.goexit-->runtime.main-->main.main-->govcd.(*Vdc).QueryWithNotEncodedParams
2019/04/11 08:11:04 GET https://192.168.1.109/api/query?&type=portgroup
2019/04/11 08:11:04 --------------------------------------------------------------------------------
2019/04/11 08:11:04 Req header:
2019/04/11 08:11:04 	X-Vcloud-Authorization: [********]
2019/04/11 08:11:04 	Accept: [application/*+xml;version=27.0]
2019/04/11 08:11:04 ################################################################################
2019/04/11 08:11:04 Response caller runtime.goexit-->runtime.main-->main.main-->govcd.(*Vdc).QueryWithNotEncodedParams-->govcd.getResult-->govcd.decodeBody
2019/04/11 08:11:04 Response status 200 OK
2019/04/11 08:11:04 ################################################################################
2019/04/11 08:11:04 Response header:
2019/04/11 08:11:04 	Date: [Thu, 11 Apr 2019 08:15:52 GMT]
2019/04/11 08:11:04 	X-Vmware-Vcloud-Request-Id: [5b5f1401-7d52-4548-b506-f3763ac4edb4]
2019/04/11 08:11:04 	X-Vmware-Vcloud-Ceip-Id: [af7af98a-f827-45c4-8bc4-dff4b81768b8]
2019/04/11 08:11:04 	Cache-Control: [no-store, must-revalidate]
2019/04/11 08:11:04 	Content-Length: [2586]
2019/04/11 08:11:04 	X-Vcloud-Authorization: [********]
2019/04/11 08:11:04 	Content-Type: [application/vnd.vmware.vcloud.query.records+xml;version=27.0]
2019/04/11 08:11:04 	X-Vmware-Vcloud-Request-Execution-Time: [11]
2019/04/11 08:11:04 	Vary: [Accept-Encoding, User-Agent]
2019/04/11 08:11:04 Response text: [2586] <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<QueryResultRecords xmlns="http://www.vmware.com/vcloud/v1.5" xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:vssd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData" xmlns:common="http://schemas.dmtf.org/wbem/wscim/1/common" xmlns:rasd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData" xmlns:vmw="http://www.vmware.com/schema/ovf" xmlns:ovfenv="http://schemas.dmtf.org/ovf/environment/1" xmlns:vmext="http://www.vmware.com/vcloud/extension/v1.5" xmlns:ns9="http://www.vmware.com/vcloud/versions" name="portgroup" page="1" pageSize="25" total="5" href="https://192.168.1.109/api/query?type=portgroup&amp;page=1&amp;pageSize=25&amp;format=records" type="application/vnd.vmware.vcloud.query.records+xml">
    <Link rel="alternate" href="https://192.168.1.109/api/query?type=portgroup&amp;page=1&amp;pageSize=25&amp;format=idrecords" type="application/vnd.vmware.vcloud.query.idrecords+xml"/>
    <PortgroupRecord isVCEnabled="true" moref="dvportgroup-17" name="Management" networkName="--" portgroupType="DV_PORTGROUP" scopeType="-1" vc="https://192.168.1.109/api/admin/extension/vimServer/4c7d5352-772c-4488-919e-7328b9767204" vcName="vC1"/>
    <PortgroupRecord isVCEnabled="true" moref="dvportgroup-18" name="VM Network" network="https://192.168.1.109/api/network/3eef8448-dc9f-4bb9-b1c2-6d599dec7ee2" networkName="vb_externalNet2" portgroupType="DV_PORTGROUP" scopeType="1" vc="https://192.168.1.109/api/admin/extension/vimServer/4c7d5352-772c-4488-919e-7328b9767204" vcName="vC1"/>
    <PortgroupRecord isVCEnabled="true" moref="dvportgroup-22" name="vxw-vmknicPg-dvs-15-0-cc6fc006-140c-4579-90bc-9c647cfcbd95" networkName="--" portgroupType="DV_PORTGROUP" scopeType="-1" vc="https://192.168.1.109/api/admin/extension/vimServer/4c7d5352-772c-4488-919e-7328b9767204" vcName="vC1"/>
    <PortgroupRecord isVCEnabled="true" moref="dvportgroup-33" name="vxw-dvs-15-virtualwire-1-sid-5000-dvs.VCDVSmyOrgNet1-fb694368-d6fe-410f-87c7-36b" network="https://192.168.1.109/api/network/fb694368-d6fe-410f-87c7-36b6fd7a6c2f" networkName="myOrgNet1" portgroupType="DV_PORTGROUP" scopeType="2" vc="https://192.168.1.109/api/admin/extension/vimServer/4c7d5352-772c-4488-919e-7328b9767204" vcName="vC1"/>
    <PortgroupRecord isVCEnabled="true" moref="network-14" name="VM Network" networkName="--" portgroupType="NETWORK" scopeType="-1" vc="https://192.168.1.109/api/admin/extension/vimServer/4c7d5352-772c-4488-919e-7328b9767204" vcName="vC1"/>
</QueryResultRecords>
 ```